File: flutter_theoplayer_sdk_android/android/src/main/kotlin/com/theoplayer/theoplayer/transformers/SourceTransformer.kt
package com.theoplayer.theoplayer.transformers

import com.theoplayer.android.api.source.SourceDescription
import com.theoplayer.android.api.source.TypedSource
import com.theoplayer.android.api.source.drm.DRMConfiguration
import com.theoplayer.android.api.source.drm.FairPlayKeySystemConfiguration
import com.theoplayer.android.api.source.drm.KeySystemConfiguration

typealias FlutterSourceDescription = com.theoplayer.theoplayer.pigeon.SourceDescription
typealias FlutterTypedSource = com.theoplayer.theoplayer.pigeon.TypedSource
typealias FlutterDRMConfiguration = com.theoplayer.theoplayer.pigeon.DRMConfiguration
typealias FlutterWidevineDRMConfiguration = com.theoplayer.theoplayer.pigeon.WidevineDRMConfiguration
typealias FlutterFairPlayDRMConfiguration = com.theoplayer.theoplayer.pigeon.FairPlayDRMConfiguration

object SourceTransformer {

    fun toFlutterSourceDescription(sourceDescription: SourceDescription?): FlutterSourceDescription? {
        if (sourceDescription == null) {
            return null
        }

        return FlutterSourceDescription(sourceDescription.sources
            .map { toFlutterTypedSource(it) }
        )
    }

    fun toFlutterTypedSource(typedSource: TypedSource?): FlutterTypedSource? {
        if (typedSource == null) {
            return null
        }

        val drm = typedSource.drm?.let {
            toFlutterDRMConfiguration(it)
        }

        return FlutterTypedSource(typedSource.src, drm)
    }

    fun toFlutterDRMConfiguration(drmConfiguration: DRMConfiguration): FlutterDRMConfiguration {
        var flutterWidevineConfig: FlutterWidevineDRMConfiguration? = null
        drmConfiguration.widevine?.let {
            flutterWidevineConfig = FlutterWidevineDRMConfiguration(it.licenseAcquisitionURL, it.headers)
        }

        var flutterFairplayConfig: FlutterFairPlayDRMConfiguration? = null
        drmConfiguration.fairplay?.let {
            flutterFairplayConfig = FlutterFairPlayDRMConfiguration(it.licenseAcquisitionURL, it.certificateURL, it.headers)
        }

        var integrationParamaters : MutableMap<String, String>? = null
        drmConfiguration.integrationParameters?.let {
            integrationParamaters = mutableMapOf()
            it.forEach { entry ->
                val value = entry.value
                if (value is String) {
                    integrationParamaters!![entry.key] = value
                }
            }
        }

        return FlutterDRMConfiguration(
            widevine = flutterWidevineConfig,
            fairplay = flutterFairplayConfig,
            customIntegrationId = drmConfiguration.customIntegrationId,
            integrationParameters = integrationParamaters?.toMap()
        )
    }

    fun toSourceDescription(flutterSourceDescription: FlutterSourceDescription?): SourceDescription? {
        if (flutterSourceDescription == null) {
            return null
        }

        return SourceDescription.Builder(
            *flutterSourceDescription.sources
                .map { toTypedSource(it) }
                .toTypedArray())
            .build()
    }

    fun toTypedSource(flutterTypedSource: FlutterTypedSource?): TypedSource? {
        if (flutterTypedSource == null) {
            return null
        }

        val typedSourceBuilder = TypedSource.Builder(flutterTypedSource.src)

        flutterTypedSource.drm?.let {
            typedSourceBuilder.drm(toDRMConfiguration(it))
        }

        return typedSourceBuilder.build()
    }

    fun toDRMConfiguration(flutterDRMConfiguration: FlutterDRMConfiguration): DRMConfiguration {
        val drm = DRMConfiguration.Builder()

        flutterDRMConfiguration.widevine?.let { flutterWidevineConfig ->
            val widevineDRMConfiguration = KeySystemConfiguration.Builder(flutterWidevineConfig.licenseAcquisitionURL)
            flutterWidevineConfig.headers?.let {
                widevineDRMConfiguration.headers(it)
            }
            drm.widevine(widevineDRMConfiguration.build())
        }

        flutterDRMConfiguration.fairplay?.let { flutterFairplayConfig ->
            val fairplayDRMConfiguration = FairPlayKeySystemConfiguration.Builder(flutterFairplayConfig.licenseAcquisitionURL, flutterFairplayConfig.certificateURL)
            flutterFairplayConfig.headers?.let {
                fairplayDRMConfiguration.headers(it)
            }
            drm.fairplay(fairplayDRMConfiguration.build())
        }

        flutterDRMConfiguration.customIntegrationId?.let {
            drm.customIntegrationId(it);
        }

        flutterDRMConfiguration.integrationParameters?.let {
            drm.integrationParameters(it)
        }

        return drm.build()
    }

}

File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_ios/ios/Classes/transformers/SourceTransformer.swift
//
//  SourceTransformer.swift
//  flutter_theoplayer_sdk_ios
//
//  Created by Hovig on 16/10/2023.
//

import Foundation
import THEOplayerSDK
import THEOplayerTHEOliveIntegration

struct SourceTransformer {
    static func toFlutterSourceDescription(source: THEOplayerSDK.SourceDescription?) -> SourceDescription? {
        guard let source = source else {
            return nil
        }
        
        return SourceDescription(
            sources: source.sources.map({ typedSource in
                toFlutterTypedSource(typedSource: typedSource)
            })
        )
    }
    
    static func toFlutterTypedSource(typedSource: THEOplayerSDK.TypedSource?) -> PigeonTypedSource? {
        guard let typedSource = typedSource else {
            return nil
        }
        
        var flutterDRMConfiguration: DRMConfiguration?
        if let drmConfiguration = typedSource.drm {
            flutterDRMConfiguration = toFlutterDRMConfiguration(drmConfiguration: drmConfiguration)
        }
        
        return PigeonTypedSource(src: typedSource.src, drm: flutterDRMConfiguration, playbackPipeline: PlaybackPipeline.legacy, headers: typedSource.headers)
    }
    
    
    static func toFlutterDRMConfiguration(drmConfiguration: THEOplayerSDK.DRMConfiguration) -> DRMConfiguration {
        var flutterFairplayConfig: FairPlayDRMConfiguration?
        if let fairPlayDRMConfiguration = drmConfiguration as? THEOplayerSDK.FairPlayDRMConfiguration {
            flutterFairplayConfig = FairPlayDRMConfiguration(
                licenseAcquisitionURL: fairPlayDRMConfiguration.fairplay.licenseAcquisitionURL?.absoluteString ?? "",
                certificateURL: fairPlayDRMConfiguration.fairplay.certificateURL?.absoluteString ?? "",
                headers: fairPlayDRMConfiguration.fairplay.headers)
        }

        var flutterWidevineConfig: WidevineDRMConfiguration?
        if let widevineDRMConfiguration = drmConfiguration as? THEOplayerSDK.WidevineDRMConfiguration {
            flutterWidevineConfig = WidevineDRMConfiguration(
                licenseAcquisitionURL: widevineDRMConfiguration.widevine.licenseAcquisitionURL?.absoluteString ?? "",
                headers: widevineDRMConfiguration.widevine.headers)
        }
        
        
        var convertedIntegrationParameters: [String: String]? = nil
        if let integrationParameters = drmConfiguration.integrationParameters as? [String:String] {
            convertedIntegrationParameters = integrationParameters
        } else {
            print("Failed to convert integrationParameters!: ", drmConfiguration.integrationParameters);
        }
        
        return DRMConfiguration(widevine: flutterWidevineConfig, fairplay: flutterFairplayConfig, customIntegrationId: drmConfiguration.customIntegrationId, integrationParameters: convertedIntegrationParameters)
    }
    
    
    static func toSourceDescription(source: SourceDescription?) -> THEOplayerSDK.SourceDescription? {
        guard let source = source else {
            return nil
        }
        
        return THEOplayerSDK.SourceDescription(
            sources: source.sources
                .map({ typedSource in
                    toTypedSource(typedSource: typedSource)
                })
                .compactMap({ $0 })
        )
    }
    
    static func toTypedSource(typedSource: PigeonTypedSource?) -> THEOplayerSDK.TypedSource? {
        guard let typedSource = typedSource else {
            return nil
        }
        
        switch typedSource.integration {
            case .theolive:
                return TheoLiveSource(channelId: typedSource.src)
            default:
                let drm = toDRMConfiguration(flutterDRMConfiguration: typedSource.drm)
                
                return THEOplayerSDK.TypedSource(
                    src: typedSource.src,
                    type: "",
                    drm: drm,
                    headers: cleanOptionalHashMap(typedSource.headers)
                )
        }
    }
    
    static func cleanOptionalHashMap(_ optionalHashMap: [String?: String?]?) -> [String: String]? {
        let cleaned: [String: String]? = optionalHashMap?.compactMap { (key, value) -> (String, String)? in
            if let key = key, let value = value {
                return (key, value)
            } else {
                return nil
            }
        }.reduce(into: [String: String]()) { result, pair in
            result[pair.0] = pair.1
        }
        
        return cleaned
    }
    
    static func toDRMConfiguration(flutterDRMConfiguration: DRMConfiguration?) -> THEOplayerSDK.DRMConfiguration? {
        guard let flutterDRMConfiguration = flutterDRMConfiguration else {
            return nil
        }
        
        if let widevine = flutterDRMConfiguration.widevine {
            var headers: Array<[String : String]> = []
            widevine.headers?.forEach({ header in
                if let key = header.key,
                   let value = header.value {
                    headers.append([key: value])
                }
            })
            
            if let customIntegration = flutterDRMConfiguration.customIntegrationId {
                var integrationParameters: [String : String] = [:]
                flutterDRMConfiguration.integrationParameters?.forEach({ header in
                    if let key = header.key,
                       let value = header.value {
                        integrationParameters[key] = value
                    }
                })
                
                return THEOplayerSDK.WidevineDRMConfiguration(
                    customIntegrationId: customIntegration,
                    licenseAcquisitionURL: widevine.licenseAcquisitionURL,
                    integrationParameters: integrationParameters
                )
            } else {
                return THEOplayerSDK.WidevineDRMConfiguration(
                    licenseAcquisitionURL: widevine.licenseAcquisitionURL,
                    headers: headers
                )
            }

        }
        
        if let fairplay = flutterDRMConfiguration.fairplay {
            var headers: Array<[String : String]> = []
            fairplay.headers?.forEach({ header in
                if let key = header.key,
                   let value = header.value {
                    headers.append([key: value])
                }
            })
            
            if let customIntegration = flutterDRMConfiguration.customIntegrationId {
                var integrationParameters: [String : String] = [:]
                flutterDRMConfiguration.integrationParameters?.forEach({ header in
                    if let key = header.key,
                       let value = header.value {
                        integrationParameters[key] = value
                    }
                })
                return THEOplayerSDK.FairPlayDRMConfiguration(
                    customIntegrationId: customIntegration,
                    licenseAcquisitionURL: fairplay.licenseAcquisitionURL,
                    certificateURL: fairplay.certificateURL,
                    headers: headers,
                    integrationParameters: integrationParameters
                )
            } else {
                return THEOplayerSDK.FairPlayDRMConfiguration(
                    licenseAcquisitionURL: fairplay.licenseAcquisitionURL,
                    certificateURL: fairplay.certificateURL,
                    headers: headers
                )
            }
        }
        
        return nil
        
    }
}
